src/Adapter/CustomerService/CommandHandler/ReplyToCustomerThreadHandler.php

```diff
@@ -0,0 +1,174 @@
+use PrestaShop\PrestaShop\Core\Domain\CustomerService\CommandHandler\ReplyToCustomerThreadHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\CustomerService\Exception\CustomerServiceException;
+use PrestaShop\PrestaShop\Core\Domain\CustomerService\ValueObject\CustomerThreadStatus;
+use ShopUrl;
+use Symfony\Component\Translation\TranslatorInterface;
+use Tools;
+use Validate;
+
+/**
+ * @internal
+ */
+final class ReplyToCustomerThreadHandler implements ReplyToCustomerThreadHandlerInterface
+{
+    /**
+     * @var Context
+     */
+    private $context;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @param Context $context
+     */
+    public function __construct(Context $context)
+    {
+        $this->context = $context;
+        $this->translator = $context->getTranslator();
+    }
+
+    /**
+     * @param ReplyToCustomerThreadCommand $command
+     */
+    public function handle(ReplyToCustomerThreadCommand $command)
+    {
+        $customerThread = new CustomerThread(
+            $command->getCustomerThreadId()->getValue()
+        );
+
+        ShopUrl::cacheMainDomainForShop((int) $customerThread->id_shop);
+
+        $this->createCustomerMessage(
+            $customerThread,
+            $command->getReplyMessage()
+        );
+
+        $replyWasSent = $this->sendReplyEmail($customerThread, $command);
+
+        if ($replyWasSent) {
+            $customerThread->status = CustomerThreadStatus::CLOSED;
+            $customerThread->update();
+        }
+    }
+
+    /**
+     * @param CustomerThread $customerThread
+     * @param string $replyMessage
+     *
+     * @return CustomerMessage
+     */
+    private function createCustomerMessage(CustomerThread $customerThread, $replyMessage)
+    {
+        $customerMessage = new CustomerMessage();
+        $customerMessage->id_employee = (int) $this->context->employee->id;
+        $customerMessage->id_customer_thread = $customerThread->id;
+        $customerMessage->ip_address = (int) ip2long(Tools::getRemoteAddr());
+        $customerMessage->message = $replyMessage;
+
+        if (false === $customerMessage->validateField('message', $customerMessage->message)) {
+            throw new CustomerServiceException('Invalid reply message');
+        }
+
+        if (false === $customerMessage->add()) {
+            throw new CustomerServiceException('Failed to add customer message');
+        }
+
+        return $customerMessage;
+    }
+
+    /**
+     * @param CustomerThread $customerThread
+     * @param ReplyToCustomerThreadCommand $command
+     *
+     * @return bool
+     */
+    private function sendReplyEmail(CustomerThread $customerThread, ReplyToCustomerThreadCommand $command)
+    {
+        $customer = new Customer($customerThread->id_customer);
+
+        $params = [
+            '{reply}' => Tools::nl2br($command->getReplyMessage()),
```

Comment:
Should we add another layer of security 🤔like escaping all of this

Reply:
Makes sense, I've used `CustomerMessage` to access reply instead of command as it has `isCleanHtml` constraint.